app/src/main/java/com/tatar/currencify/presentation/viewmodel/ViewModelFactory.kt

```diff
@@ -0,0 +1,37 @@
+package com.tatar.currencify.presentation.viewmodel
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.ViewModelProvider
+import com.tatar.currencify.dagger.app.scope.AppScope
+import javax.inject.Inject
+import javax.inject.Provider
+
+@AppScope
+class ViewModelFactory @Inject constructor(
+    private val providerMap: @JvmSuppressWildcards MutableMap<Class<out ViewModel>, Provider<ViewModel>>
+) : ViewModelProvider.Factory {
+
+    override fun <T : ViewModel> create(modelClass: Class<T>): T {
+        var provider: Provider<out ViewModel>? = providerMap[modelClass]
+
+        if (provider == null) {
+            for ((key, value) in providerMap) {
+                if (modelClass.isAssignableFrom(key)) {
+                    provider = value
+                    break
+                }
+            }
+        }
+
+        if (provider == null) {
```

Comment:
Why not else here? so you would get rid of an if that checks same thing?

Reply:
Also, I'd prefer to put this throw in the top of function as a quick return/exit. What was the reason you put here after?

Reply:
In Swift we have one thing called guard so you would do something similar like below. So, you wouldn't need any indentation, basically code would be clearer. Does Kotlin have something similar?

Reply:
This class is a weird one. The reason for the weirdness is injecting Android's `ViewModel`s via Dagger 2. It is a long story but the story is the reason for implementing this class in that way.